src/test/kotlin/org/rust/cargo/toolchain/CargoProjectResolveTestCase.kt

```diff
@@ -9,31 +9,50 @@ import org.assertj.core.api.Assertions.assertThat
 import org.rust.cargo.RustWithToolchainTestCaseBase
 import org.rust.cargo.project.settings.toolchain
 import org.rust.cargo.project.workspace.CargoProjectWorkspace
-import org.rust.cargo.project.workspace.impl.CargoProjectWorkspaceImpl
-import org.rust.cargo.util.cargoProject
+import org.rust.cargo.project.workspace.CargoProjectWorkspaceListener
+import org.rust.cargo.project.workspace.CargoProjectWorkspaceListener.UpdateResult
+import org.rust.cargo.util.getComponentOrThrow
 import org.rust.cargo.util.getServiceOrThrow
 
 class CargoProjectResolveTestCase : RustWithToolchainTestCaseBase() {
     override val dataPath: String = "src/test/resources/org/rust/cargo/toolchain/fixtures"
 
     fun testResolveExternalLibrary() = withProject("external_library") {
-        updateCargoMetadata()
-        val reference = extractReference("src/main.rs")
-        assertThat(reference.resolve()).isNotNull()
+        module.messageBus
+            .connect()
+            .subscribe(
+                CargoProjectWorkspaceListener.Topics.UPDATES,
+                object: CargoProjectWorkspaceListener {
+                    override fun onWorkspaceUpdateCompleted(r: UpdateResult) {
+                        assertThat(r is UpdateResult.Ok)
+
+                        val reference = extractReference("src/main.rs")
+                        assertThat(reference.resolve()).isNotNull()
+                    }
+                })
+
+        updateCargoProject()
     }
 
     fun testResolveLocalPackage() = withProject("local_package") {
-        updateCargoMetadata()
-        val reference = extractReference("src/main.rs")
-        assertThat(reference.resolve()).isNotNull()
+        module.messageBus
+            .connect()
+            .subscribe(
+                CargoProjectWorkspaceListener.Topics.UPDATES,
+                object: CargoProjectWorkspaceListener {
+                    override fun onWorkspaceUpdateCompleted(r: UpdateResult) {
+                        assertThat(r is UpdateResult.Ok)
+
+                        val reference = extractReference("src/main.rs")
+                        assertThat(reference.resolve()).isNotNull()
+                    }
+                })
+
```

Comment:
Async tests are tricky. If the message is never delivered, this code never gets executed and the test pass happily. 

In essence, what we want to test here is that the message is actually delivered. This is a hard problem.